### PR TITLE
Add a restore and add pkg commands interactive flag

### PR DESCRIPTION
**Pull request by nkolev92**
This change is dependent on https://github.com/NuGet/NuGet.Client/pull/2288. 
Ideally it should be merged when an insertion that contains the above PR happens. 

Fixes https://github.com/NuGet/Home/issues/7017

Part of https://github.com/NuGet/Home/issues/6642. 

Per https://github.com/NuGet/Home/wiki/NuGet-cross-plat-authentication-plugin, we agreed to use the
`--interactive` option in the dotnet restore and add package ref commands to allow the action to be blocked. 

Let me know if I should be targetting a different branch. The change should make it 2.1.4xx in the end. 

//cc
@livarcocc @nguerrera @dsplaisted @peterhuene @wli3

### CODE REVIEW

**Comment by nkolev92** [comment]
@livarcocc @nguerrera @dsplaisted @peterhuene @wli3 
Can I get some eyes here please?

File: src/dotnet/commands/dotnet-add/dotnet-add-package/AddPackageParser.cs [hunk 1]
```diff
@@ -47,7 +47,9 @@ public static Command AddPackage()
                               LocalizableStrings.CmdPackageDirectoryDescription,
                               Accept.ExactlyOneArgument()
                                     .With(name: LocalizableStrings.CmdPackageDirectory)
-                                    .ForwardAsSingle(o => $"--package-directory {o.Arguments.Single()}")));
+                                    .ForwardAsSingle(o => $"--package-directory {o.Arguments.Single()}")),
+                Create.Option("--interactive",
```
Comment (wli3) [inline src/dotnet/commands/dotnet-add/dotnet-add-package/AddPackageParser.cs]: this map to a nuget.xplat interface right?

Comment (wli3) [reply src/dotnet/commands/dotnet-add/dotnet-add-package/AddPackageParser.cs]: i think there should be some code to pass it down.

**Comment by wli3** [comment]
@KathleenDollard for user facing string review

File: src/dotnet/commands/dotnet-add/dotnet-add-package/AddPackageParser.cs (hunk 1 above)
Comment (wli3) [reply src/dotnet/commands/dotnet-add/dotnet-add-package/AddPackageParser.cs]: It should be something like this line https://github.com/nkolev92/cli/blob/c7cde12e2def4f962db8d72ee6a25e29b424f49f/src/dotnet/commands/dotnet-add/dotnet-add-package/Program.cs#L148

It is a pass through, so I don't think automatic tests are necessary, but please verify by run these command by hand

**Comment by wli3** [comment]
and.. i don't know about the failing tests. No dependency changed, why package downgrade

**Comment by wli3** [comment]
@dotnet-bot test this please

File: src/dotnet/commands/dotnet-add/dotnet-add-package/LocalizableStrings.resx [hunk 2]
```diff
@@ -165,4 +165,7 @@
   <data name="CmdDGFileIOException" xml:space="preserve">
     <value>Unable to generate a temporary file for project '{0}'. Cannot add package reference. Please clear the temp directory and try again.</value>
   </data>
+  <data name="CmdInteractiveRestoreDescription" xml:space="preserve">
+    <value>Allow the command to block and require manual action to proceed.</value>
```
Comment (dsplaisted) [inline src/dotnet/commands/dotnet-add/dotnet-add-package/LocalizableStrings.resx]: I'd suggest rewording.  Maybe PMs or doc writers can come up with something better, but I'd suggest something like:

> Allow the command to stop and wait for user input or action if required (for example in order to complete authentication).

File: src/dotnet/commands/dotnet-restore/LocalizableStrings.resx [hunk 3]
```diff
@@ -169,4 +169,7 @@
     <value>Force all dependencies to be resolved even if the last restore was successful.
 This is equivalent to deleting project.assets.json.</value>
   </data>
+  <data name="CmdInteractiveRestoreOptionDescription" xml:space="preserve">
+    <value>Allow restore to block and require manual action to proceed.</value>
```
Comment (dsplaisted) [inline src/dotnet/commands/dotnet-restore/LocalizableStrings.resx]: If the previous message is reworded, this one should be reworded too.

File: src/dotnet/commands/dotnet-add/dotnet-add-package/LocalizableStrings.resx (hunk 2 above)
Comment (KathleenDollard) [reply src/dotnet/commands/dotnet-add/dotnet-add-package/LocalizableStrings.resx]: I tend to 

> The command may stop and wait for user input or action (for example to complete authentication)

File: src/dotnet/commands/dotnet-restore/LocalizableStrings.resx (hunk 3 above)
Comment (KathleenDollard) [reply src/dotnet/commands/dotnet-restore/LocalizableStrings.resx]: I would suggest exactly the same message - but I'm assuming they are both --interactive options.

File: src/dotnet/commands/dotnet-add/dotnet-add-package/LocalizableStrings.resx (hunk 2 above)
Comment (KathleenDollard) [reply src/dotnet/commands/dotnet-add/dotnet-add-package/LocalizableStrings.resx]: @dsplaisted and I agreed on

> Allow the command to stop and wait for user input or action (for example to complete authentication)

@mairaw Should this be `Allow` or `Allows`

File: src/dotnet/commands/dotnet-restore/RestoreCommandParser.cs [hunk 4]
```diff
@@ -94,8 +94,13 @@ private static Option[] ImplicitRestoreOptions(bool showHelp = false, bool useSh
                     useShortOptions ? "-f|--force" : "--force",
                     LocalizableStrings.CmdForceRestoreOptionDescription,
                     Accept.NoArguments()
-                          .ForwardAs("-property:RestoreForce=true"))
+                          .ForwardAs("-property:RestoreForce=true")),
+                 Create.Option(
```
Comment (dsplaisted) [inline src/dotnet/commands/dotnet-restore/RestoreCommandParser.cs]: Does the fact that this is in an ImplicitRestoreOptions method mean that this option applies to operations such as `dotnet build` or `dotnet publish`?  I think we had said that it would not.

File: src/dotnet/commands/dotnet-restore/LocalizableStrings.resx (hunk 3 above)
Comment (dsplaisted) [reply src/dotnet/commands/dotnet-restore/LocalizableStrings.resx]: Discussed with Kathleen, let's use the same message for both (ie it's OK to say "the command" instead of "restore") in both cases.

File: src/dotnet/commands/dotnet-add/dotnet-add-package/LocalizableStrings.resx (hunk 2 above)
Comment (mairaw) [reply src/dotnet/commands/dotnet-add/dotnet-add-package/LocalizableStrings.resx]: I'd like to change the --help messages to use 3rd person throughout. So Allows...

File: src/dotnet/commands/dotnet-restore/RestoreCommandParser.cs (hunk 4 above)
Comment (nkolev92) [reply src/dotnet/commands/dotnet-restore/RestoreCommandParser.cs]: @dsplaisted 
good point...double checked the meeting notes, that was the agreement. 

I'm not sure how make a command not propagate though...can you point me to the code?

File: src/dotnet/commands/dotnet-add/dotnet-add-package/AddPackageParser.cs (hunk 1 above)
Comment (nkolev92) [reply src/dotnet/commands/dotnet-add/dotnet-add-package/AddPackageParser.cs]: This piece of code suggests I don't need to do it. 

https://github.com/nkolev92/cli/blob/c7cde12e2def4f962db8d72ee6a25e29b424f49f/src/dotnet/commands/dotnet-add/dotnet-add-package/Program.cs#L144-L146

I'll verify by manually building the cli + nuget.

File: src/dotnet/commands/dotnet-restore/RestoreCommandParser.cs (hunk 4 above)
Comment (dsplaisted) [reply src/dotnet/commands/dotnet-restore/RestoreCommandParser.cs]: I'm not familiar myself with how the options/parsing code works.  @wli3 may be able to help.

Comment (peterhuene) [reply src/dotnet/commands/dotnet-restore/RestoreCommandParser.cs]: If this is meant only for the `restore` command, then it probably belongs in the Concat call [here](https://github.com/nkolev92/cli/blob/c834c8cd0b147e5232460174a1d6e89310b7a9e1/src/dotnet/commands/dotnet-restore/RestoreCommandParser.cs#L27).

Comment (nkolev92) [reply src/dotnet/commands/dotnet-restore/RestoreCommandParser.cs]: Thanks @peterhuene, that worked. 

It's a bit difficult to understand how the differentiation is being made between those 2 before I saw the CreateWithRestoreOptions helper.

Comment (peterhuene) [reply src/dotnet/commands/dotnet-restore/RestoreCommandParser.cs]: Yeah, this could use a little refactoring for comprehensibility's sake.

File: src/dotnet/commands/dotnet-add/dotnet-add-package/AddPackageParser.cs (hunk 1 above)
Comment (nkolev92) [reply src/dotnet/commands/dotnet-add/dotnet-add-package/AddPackageParser.cs]: @wli3 
I added a forwarding call.

**Comment by nkolev92** [comment]
🔔 
@dsplaisted @wli3 @peterhuene 
I've addressed the feedback.

File: src/dotnet/commands/dotnet-add/dotnet-add-package/LocalizableStrings.resx [hunk 5]
```diff
@@ -165,4 +165,7 @@
   <data name="CmdDGFileIOException" xml:space="preserve">
     <value>Unable to generate a temporary file for project '{0}'. Cannot add package reference. Please clear the temp directory and try again.</value>
   </data>
+  <data name="CmdInteractiveRestoreDescription" xml:space="preserve">
+    <value>Allows the command to stop and wait for user input or action (for example to complete authentication)</value>
```
Comment (peterhuene) [inline src/dotnet/commands/dotnet-add/dotnet-add-package/LocalizableStrings.resx]: Nit: for conformity with our other option descriptions, could you add a period on the end of this string?

**Comment by nkolev92** [comment]
@dsplaisted @peterhuene @wli3 

The way NuGet plugins work in dotnet.exe is that NuGet runs "dotnet exec dll". 

To make sure that the correct runtime is used and to make it executes in the first place, we look for the DOTNET_HOST_PATH env var.  
Code [here](https://github.com/NuGet/NuGet.Client/blob/b14b0500100420e4da687206b4b1cfb08a0a437d/src/NuGet.Core/NuGet.Protocol/Plugins/PluginFactory.cs#L159-L170)

Now that's fine with restore (because it runs in the msbuild assembly), and the way it's [launched](https://github.com/dotnet/cli/blob/master/src/Microsoft.DotNet.Cli.Utils/MSBuildForwardingAppWithoutLogging.cs#L27).

add package currently has problems because that var is not set. 

I know I can solve this by forwarding the DOTNET_HOST_PATH to the add package command (I have tested that and it works), by adding it [here](https://github.com/dotnet/cli/blob/be8ed9558be389b3e40a5758d157019397a6290e/src/dotnet/commands/dotnet-nuget/Program.cs#L37). 

Any other options? 
I read something about DOTNET_ROOT but that doesn't seem to work.

**Comment by wli3** [comment]
Sorry for the delay. @nkolev92 

DOTNET_ROOT is not the right one. `DOTNET_ROOT` is part of runtime's(framework **dependent** apphost's) logic to find the framework. So an application(nuget.xplat in this case) should not operate on this level of abstraction. The problem you mention is a **different** topic. The full context is here https://github.com/dotnet/designs-microsoft/pull/34/files . In short, we still want to use DOTNET_ROOT moving forward.

@dsplaisted see if you agree with me. I think nuget.xplat should only relay on OS to find dotnet.exe and muxer to pick the right runtime. However, if it is important that "use the dotnet that called me" passing down DOTNET_HOST_PATH or edit PATH should be correct.

**Comment by nkolev92** [comment]
@wli3 
Thanks for the explanation. 

It's important that "the dotnet that called me" is used. 
Otherwise we might get weird incompatibility with people that have multiple instances. 

I'd like to pass down the DOTNET_HOST_PATH to the NuGet.XPLAT once @dsplaisted 

Should I do it in this PR?

**Comment by dsplaisted** [comment]
@nkolev92 @wli3 I think setting DOTNET_HOST_PATH for the add package command is fine.

You could still have issues if you have plugins in the UserProfile that depend on a runtime that isn't available in your local copy of dotnet.  However, I think the solution there would be to set the NUGET_PLUGIN_PATHS environment variable so you aren't using the UserProfile plugins.

**Comment by nkolev92** [comment]
@dsplaisted 
Yes that's correct.
It's a problem we are not trying to solve, because even having tfm based plugins doesn't work. 
We're taking the simple approach, will optimize if we need to in the future.

**Comment by nkolev92** [comment]
🔔 
@wli3 @dsplaisted 
I have updated the PR.

**Comment by nkolev92** [comment]
@dotnet-bot Test Ubuntu x64 Release Build

File: src/dotnet/commands/dotnet-nuget/Program.cs [hunk 6]
```diff
@@ -35,9 +35,14 @@ private class NuGetCommandRunner : ICommandRunner
             public int Run(string[] args)
             {
                 var nugetApp = new NuGetForwardingApp(args);
-
+                nugetApp.WithEnvironmentVariable("DOTNET_HOST_PATH", GetDotnetPath());
```
Comment (wli3) [inline src/dotnet/commands/dotnet-nuget/Program.cs]: ping again @dsplaisted does it look good to you?

**Comment by wli3** [comment]
@livarcocc need your approval for 2.1.4xx

File: src/dotnet/commands/dotnet-nuget/Program.cs (hunk 6 above)
Comment (dsplaisted) [reply src/dotnet/commands/dotnet-nuget/Program.cs]: @wli3 Yes, assuming `GetDotnetPath()` returns the value we want here.

**Comment by wli3** [comment]
@nkolev92 once Livar approved, whoever see it first merge it
